File: apps/android/app/src/main/java/com/kungsbackacarcommunity/app/chatchannels/FirebaseCommunityChatRepository.kt
package com.kungsbackacarcommunity.app.chatchannels

import android.content.Context
import com.google.firebase.FirebaseApp
import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.firestore.FirebaseFirestoreException
import com.google.firebase.firestore.Query
import com.google.firebase.functions.FirebaseFunctions
import com.kungsbackacarcommunity.app.blocking.BlockVisibility
import com.kungsbackacarcommunity.app.blocking.BlockVisibilityRepository
import com.kungsbackacarcommunity.app.blocking.FirebaseBlockVisibilityRepository
import com.kungsbackacarcommunity.app.profile.FirebaseLiveProfileRepository
import com.kungsbackacarcommunity.app.profile.LiveProfileRepository
import com.kungsbackacarcommunity.app.profile.LiveProfiles
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.channels.awaitClose
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.callbackFlow
import kotlinx.coroutines.flow.combine
import kotlinx.coroutines.flow.flatMapLatest
import kotlinx.coroutines.flow.flowOf
import kotlinx.coroutines.flow.map

/**
 * [CommunityChatRepository] backed by the member-readable
 * `communityChat/global/messages` listener plus the member-gated
 * `communityChat-*` callables (europe-west1). Guarded ([createIfAvailable]) so a
 * config-less build gets a null repository and the screen renders a placeholder.
 *
 * The newest-window listener is bounded (createdAt descending, capped at
 * [CHANNEL_MESSAGES_PAGE_SIZE]) so it never syncs the whole channel; older pages
 * come from `communityChat-list`. Unread combines a newest-message listener with
 * the owner-readable `userPrivate/{uid}.communityChatLastReadAt` marker.
 *
 * BLOCKING: both the live window and the unread dot are filtered against the
 * caller's mutual-hidden set ([BlockVisibilityRepository]) so a blocked pair
 * never sees each other's messages, whichever side blocked. Older pages are
 * filtered SERVER-side by `communityChat-list`; the live window cannot be,
 * because a Firestore rule cannot filter a list query per document — see
 * [BlockVisibility] for the full enforcement map.
 *
 * LIVE PROFILES: a message carries the sender's name/avatar as they were at post
 * time, stamped on by `communityChat-post` and never rewritten. Both the live
 * window and older pages are overlaid with the sender's current `users/{uid}`
 * profile ([LiveProfileRepository], [ChannelThread.hydrate]) so a member who
 * changes their avatar changes it on their whole history — de-duplicated by
 * sender, never a read per message.
 */
class FirebaseCommunityChatRepository private constructor(
    private val firestore: FirebaseFirestore,
    private val functions: FirebaseFunctions,
    private val blockVisibility: BlockVisibilityRepository,
    private val liveProfiles: LiveProfileRepository,
) : CommunityChatRepository {

    private fun messagesQuery(limit: Long): Query =
        firestore
            .collection(COMMUNITY)
            .document(CHANNEL_ID)
            .collection(MESSAGES)
            .orderBy(CREATED_AT, Query.Direction.DESCENDING)
            .limit(limit)

    @OptIn(ExperimentalCoroutinesApi::class)
    override fun observeMessages(): Flow<ChannelMessagesState> =
        combine(observeRawMessages(), blockVisibility.observeHiddenUids()) { state, hidden ->
            // The hidden set is loaded ONCE per session (one document listener),
            // never per message, so the filter costs no reads at all.
            when (state) {
                is ChannelMessagesState.Loaded ->
                    ChannelMessagesState.Loaded(
                        BlockVisibility.filterHiddenAuthors(state.messages, hidden) { it.senderUid },
                    )
                ChannelMessagesState.Loading -> state
            }
        }
            // Overlay each sender's CURRENT profile onto the copy stamped on the
            // message at post time (ChannelThread.hydrate explains the decision).
            //
            // De-duplicated by sender BEFORE the read, so a full window costs a
            // read per distinct sender rather than per message. Runs after the
            // block filter, so a hidden sender is never paid for.
            .flatMapLatest { state ->
                if (state !is ChannelMessagesState.Loaded) return@flatMapLatest flowOf(state)
                val uids = LiveProfiles.uidsOf(state.messages) { it.senderUid }
                liveProfiles.observeProfiles(uids).map { live ->
                    ChannelMessagesState.Loaded(ChannelThread.hydrate(state.messages, live))
                }
            }

    private fun observeRawMessages(): Flow<ChannelMessagesState> = callbackFlow {
        val registration =
            messagesQuery(CHANNEL_MESSAGES_PAGE_SIZE.toLong())
                .addSnapshotListener { snapshot, error ->
                    if (error != null) {
                        if ((error as? FirebaseFirestoreException)?.code ==
                            FirebaseFirestoreException.Code.PERMISSION_DENIED
                        ) {
                            // Access revoked (lost membership / blocked): emit the
                            // empty/denied state, clearing to no messages even when a
                            // cached snapshot exists, so denied history is never
                            // shown. Applies whether or not a snapshot came back (a
                            // first-load deny also clears to empty rather than
                            // staying in Loading). PERMISSION_DENIED is terminal — no
                            // retry undoes it.
                            trySend(ChannelMessagesState.Loaded(emptyList()))
                        } else if (snapshot != null) {
                            // Transient failure (UNAVAILABLE/timeout) WITH cached
                            // data: prefer the last-known messages; the SDK retries
                            // and delivers a fresh snapshot.
                            val cached =
                                snapshot.documents.mapNotNull { it.toChannelMessage() }.asReversed()
                            trySend(ChannelMessagesState.Loaded(cached))
                        }
                        // Transient failure with NO cached data: don't emit — an
                        // empty list would misrender offline/unavailable as "no
                        // messages". The SDK retries; the initial Loading holds
                        // until a real snapshot arrives.
                        return@addSnapshotListener
                    }
                    val messages =
                        snapshot?.documents?.mapNotNull { it.toChannelMessage() }
                            ?.asReversed().orEmpty()
                    trySend(ChannelMessagesState.Loaded(messages))
                }
        awaitClose { registration.remove() }
    }

    override fun observeUnread(uid: String): Flow<Boolean> {
        // The window is [UNREAD_SCAN_LIMIT], not 1, because the newest message may
        // be from a blocked party: lighting the dot for a message the user will
        // never be shown sends them into an apparently unchanged channel. Scanning
        // a few and taking the newest VISIBLE one fixes that for any realistic
        // case; the documented bound is that if ALL of the newest
        // [UNREAD_SCAN_LIMIT] messages are hidden, an older unread message does
        // not light the dot. The channel itself still shows it.
        val newestWindow: Flow<List<ChannelMessage>?> = callbackFlow {
            val registration =
                messagesQuery(UNREAD_SCAN_LIMIT).addSnapshotListener { snapshot, error ->
                    if (error != null) {
                        if ((error as? FirebaseFirestoreException)?.code ==
                            FirebaseFirestoreException.Code.PERMISSION_DENIED
                        ) {
                            // Access revoked: hard-clear unread even if a stale
                            // cached snapshot is present — never keep the dot lit
                            // for a channel the user can no longer read.
                            trySend(null)
                            return@addSnapshotListener
                        }
                        // Other (transient) error with no cached data: keep the
                        // last-known value rather than emitting a misleading
                        // no-unread. With cached data, fall through and use it.
                        if (snapshot == null) return@addSnapshotListener
                    }
                    // Newest-first (the query is createdAt DESC), so the first
                    // visible entry downstream is the newest visible message.
                    trySend(snapshot?.documents?.mapNotNull { it.toChannelMessage() }.orEmpty())
                }
            awaitClose { registration.remove() }
        }
        val newest: Flow<ChannelMessage?> =
            combine(newestWindow, blockVisibility.observeHiddenUids()) { window, hidden ->
                window?.let { BlockVisibility.newestVisible(it, hidden) { m -> m.senderUid } }
            }
        val lastReadAt: Flow<Long?> = callbackFlow {
            val registration =
                firestore
                    .collection(USER_PRIVATE)
                    .document(uid)
                    .addSnapshotListener { snapshot, error ->
                        if (error != null && snapshot == null) {
                            // Transient failure with no cached marker: keep the
                            // last-known marker rather than momentarily reading it
                            // as missing, which could wrongly flip unread.
                            return@addSnapshotListener
                        }
                        trySend(snapshot?.getTimestamp(LAST_READ_AT)?.toDate()?.time)
                    }
            awaitClose { registration.remove() }
        }
        return combine(newest, lastReadAt) { message, marker ->
            ChannelThread.hasUnread(message, uid, marker)
        }
    }

    override suspend fun post(
        text: String,
        mentionedUids: List<String>,
        clientId: String?,
    ): ChannelSendResult =
        functions.callChannel(
            POST,
            // `mentionedUids` is optional in the contract, so omit it entirely
            // when there are none rather than sending an empty array. Deduped and
            // capped here as well as in the composer, so no client state can reach
            // the server's one hard reject (> MAX_MESSAGE_MENTIONS).
            buildMap<String, Any?> {
                put("text", text.trim())
                val uids = mentionedUids.distinct().take(MAX_MESSAGE_MENTIONS)
                if (uids.isNotEmpty()) put("mentionedUids", uids)
                // Only sent when present, so an omitted key is a legacy auto-id doc
                // rather than a null the strict backend schema would reject.
                if (clientId != null) put("clientId", clientId)
            },
        ).fold(
            onSuccess = { ChannelResponseParser.parsePostSuccess(it) },
            onFailure = {
                ChannelSendResult.Failed(ChannelErrorMapper.mapSend(it.toChannelErrorCode()))
            },
        )

    override suspend fun loadOlder(before: String): ChannelOlderResult =
        functions.callChannel(LIST, mapOf("before" to before)).fold(
            onSuccess = {
                val page = ChannelResponseParser.parseMessagesPage(it)
                // An older page carries the same frozen sender copies as the live
                // window, so it needs the same overlay — otherwise scrolling back
                // would show a member's old avatar above their new one. Usually
                // free: the window's senders are already cached from the live
                // hydration above.
                val live =
                    liveProfiles.loadProfiles(
                        LiveProfiles.uidsOf(page.messages) { message -> message.senderUid },
                    )
                ChannelOlderResult.Loaded(
                    page.copy(messages = ChannelThread.hydrate(page.messages, live)),
                )
            },
            onFailure = { ChannelOlderResult.Failed },
        )

    override suspend fun markRead() {
        // Best-effort idempotent bookkeeping; a transient failure is swallowed.
        functions.callChannel(MARK_READ, emptyMap())
    }

    companion object {
        private const val COMMUNITY = "communityChat"
        private const val CHANNEL_ID = "global"
        private const val MESSAGES = "messages"
        private const val CREATED_AT = "createdAt"
        private const val USER_PRIVATE = "userPrivate"
        private const val LAST_READ_AT = "communityChatLastReadAt"
        private const val POST = "communityChat-post"
        private const val LIST = "communityChat-list"
        private const val MARK_READ = "communityChat-markRead"

        /** Newest-message window scanned for the unread dot (see observeUnread). */
        private const val UNREAD_SCAN_LIMIT = 10L

        fun createIfAvailable(context: Context): CommunityChatRepository? {
            if (FirebaseApp.getApps(context).isEmpty()) return null
            return FirebaseCommunityChatRepository(
                FirebaseFirestore.getInstance(),
                FirebaseFunctions.getInstance(CHANNEL_FUNCTIONS_REGION),
                FirebaseBlockVisibilityRepository.createOrEmpty(context),
                FirebaseLiveProfileRepository.createOrEmpty(context),
            )
        }
    }
}

File: apps/ios/KCC/Chat/FirebaseCommunityChatRepository.swift
import FirebaseAuth
import FirebaseCore
import FirebaseFirestore
import FirebaseFunctions
import Foundation

/// ``CommunityChatRepository`` backed by the member-readable
/// `communityChat/global/messages` listener plus the member-gated
/// `communityChat-*` / `chatchannels-*` callables (europe-west1) — the iOS port
/// of Android's `FirebaseCommunityChatRepository`.
///
/// The newest window is a bounded (createdAt DESCENDING, limit
/// ``channelMessagesPageSize``) live listener, block-filtered against the
/// caller's `blockVisibility/{uid}.hiddenUids` mirror; older pages come from
/// `communityChat-list` (server-side block-filtered). The client never writes the
/// message tree — `communityChat-post` stamps the authorship and denormalized
/// sender profile.
///
/// Construction is guarded (``createIfAvailable()`` returns nil without Firebase
/// config), mirroring ``FirebaseEventsRepository`` and Android's
/// `createIfAvailable`.
final class FirebaseCommunityChatRepository: CommunityChatRepository, @unchecked Sendable {
    private let firestore: Firestore
    private let functions: Functions

    private init(firestore: Firestore, functions: Functions) {
        self.firestore = firestore
        self.functions = functions
    }

    func observeMessages() -> AsyncStream<ChannelMessagesState> {
        let query = firestore
            .collection(Self.communityCollection)
            .document(Self.channelId)
            .collection(Self.messagesCollection)
            .order(by: Self.createdAtField, descending: true)
            .limit(to: channelMessagesPageSize)
        let blockDoc = currentUserId().map {
            firestore.collection(ChatFirestore.blockVisibilityCollection).document($0)
        }
        return ChannelMessagesListener.stream(messagesQuery: query, blockVisibilityDocument: blockDoc)
    }

    func loadOlder(before: String) async -> ChannelOlderResult {
        do {
            let data = try await functions.httpsCallable(Self.listCallable).call(["before": before]).data
            return .loaded(ChannelResponseParser.parseMessagesPage(data))
        } catch {
            // Transient callable failure — NOT end-of-pagination; the caller
            // offers a retry.
            return .failed
        }
    }

    func post(
        text: String,
        mentionedUids: [String],
        clientId: String?,
        replyToMessageId: String?
    ) async -> ChannelSendResult {
        // `mentionedUids` is optional in the contract, so omit it entirely when
        // empty; dedup + cap here as well as in any composer so no client state
        // reaches the server's one hard reject (> MAX_MESSAGE_MENTIONS).
        var payload: [String: Any] = ["text": text.trimmingCharacters(in: .whitespacesAndNewlines)]
        var seen = Set<String>()
        let uids = mentionedUids.filter { seen.insert($0).inserted }.prefix(channelMaxMessageMentions)
        if !uids.isEmpty { payload["mentionedUids"] = Array(uids) }
        if let clientId { payload["clientId"] = clientId }
        if let replyToMessageId { payload["replyToMessageId"] = replyToMessageId }
        do {
            let data = try await functions.httpsCallable(Self.postCallable).call(payload).data
            return ChannelResponseParser.parsePostSuccess(data)
        } catch {
            return .failed(ChannelErrorMapper.mapSend(ChatFirestore.channelErrorCode(from: error)))
        }
    }

    func markRead() async {
        // Best-effort idempotent bookkeeping; a transient failure is swallowed.
        _ = try? await functions.httpsCallable(Self.markReadCallable).call([String: Any]())
    }

    func report(messageId: String, reason: ChatReportReason) async -> ChannelReportResult {
        // `channel: "community"` fixes the scope ("global") server-side; convoyId
        // is deliberately omitted (the backend rejects it for the community
        // channel).
        let payload: [String: Any] = [
            "channel": "community",
            "messageId": messageId,
            "reason": reason.wire,
        ]
        do {
            _ = try await functions.httpsCallable(Self.reportCallable).call(payload)
            return .reported
        } catch {
            // Every failure collapses to one neutral outcome — the reporter is
            // never told which.
            return .failed
        }
    }

    func currentUserId() -> String? {
        Auth.auth().currentUser?.uid
    }

    // MARK: - Constants

    private static let communityCollection = "communityChat"
    private static let channelId = "global"
    private static let messagesCollection = "messages"
    private static let createdAtField = "createdAt"
    private static let postCallable = "communityChat-post"
    private static let listCallable = "communityChat-list"
    private static let markReadCallable = "communityChat-markRead"
    /// Shared community+convoy report callable; the payload's `channel` field
    /// selects which. NOT a `communityChat-*` export.
    private static let reportCallable = "chatchannels-reportMessage"

    // MARK: - Factory

    private static let cachedLock = NSLock()
    nonisolated(unsafe) private static var cached: FirebaseCommunityChatRepository?

    /// Returns the process-wide repository when Firebase is configured, or nil
    /// when GoogleService-Info.plist is absent (CI / config-less builds). Points
    /// the SDKs at the emulator when `FIREBASE_FIRESTORE_EMULATOR_HOST` /
    /// `FIREBASE_FUNCTIONS_EMULATOR_HOST` are set — the same seams
    /// ``FirebaseEventsRepository`` / ``KccFunctionsClient`` use.
    static func createIfAvailable() -> CommunityChatRepository? {
        guard FirebaseApp.app() != nil else { return nil }
        cachedLock.lock()
        defer { cachedLock.unlock() }
        if let cached { return cached }
        let firestore = Firestore.firestore()
        if let emulator = FirebaseEmulatorHost.parse(
            ProcessInfo.processInfo.environment["FIREBASE_FIRESTORE_EMULATOR_HOST"]
        ) {
            firestore.useEmulator(withHost: emulator.host, port: emulator.port)
        }
        let functions = Functions.functions(region: ChatFirestore.functionsRegion)
        if let emulator = FirebaseEmulatorHost.parse(
            ProcessInfo.processInfo.environment["FIREBASE_FUNCTIONS_EMULATOR_HOST"]
        ) {
            functions.useEmulator(withHost: emulator.host, port: emulator.port)
        }
        let repository = FirebaseCommunityChatRepository(firestore: firestore, functions: functions)
        cached = repository
        return repository
    }
}
